### ISSUE TITLE
Sequence Parallel / fp8 Example

### ISSUE THREAD

**Issue by jeromeku** [pytorch/torchtitan#476]
@weifengpy @wanchaol 

I've seen various mentions of `sequence parallel` (which I believe is subsumed by `tensor_parallel` per #162) training when searching through commits / PRs.  Is there an "official" example or docs for a) how to configure a training run for this parallelism strategy and b) how to compose with other forms of parallelism?  

Additionally, are there any examples / docs for using `fp8` with the various parallelism primitives?

**Comment by weifengpy** [comment]
> @weifengpy @wanchaol
> 
> I've seen various mentions of `sequence parallel` (which I believe is subsumed by `tensor_parallel` per #162) training when searching through commits / PRs. Is there an "official" example or docs for a) how to configure a training run for this parallelism strategy and b) how to compose with other forms of parallelism?
> 
> Additionally, are there any examples / docs for using `fp8` with the various parallelism primitives?

Hi @jeromeku , glad to see questions around our latest work

> Is there an "official" example or docs for a) how to configure a training run for this parallelism strategy b) how to compose with other forms of parallelism?

tensor parallel / sequence parallel tutorial by @wanchaol and @tianyu-l: https://pytorch.org/tutorials/intermediate/TP_tutorial.html

for learning by doing, you can run follwing commands
* 1D tensor parallel / sequence parallel: `CONFIG_FILE="./train_configs/debug_model.toml" NGPU=4 ./run_llama_train.sh --training.data_parallel_degree 1 --training.tensor_parallel_degree 4`
* 2D tensor parallel / sequence parallel + FSDP: `CONFIG_FILE="./train_configs/debug_model.toml" NGPU=4 ./run_llama_train.sh --training.tensor_parallel_degree 2`

> are there any examples / docs for using fp8 with the various parallelism primitives?

we are drafting a note around FSDP + float8 all-gather in 2 weeks, and 3D + float8 all-gather in 2 months

if you want a preview, you can run following commands to give it a try
* 1D TP + float8 all-gather, eager mode: `CONFIG_FILE="./train_configs/debug_model.toml" NGPU=4 ./run_llama_train.sh --training.enable_float8_linear --training.data_parallel_degree 1 --training.tensor_parallel_degree 4`
* 2D FSDP2 + TP + float8 all-gather, eager mode: `CONFIG_FILE="./train_configs/debug_model.toml" NGPU=4 ./run_llama_train.sh --training.enable_float8_linear --training.enable_fsdp_float8_all_gather --training.precompute_float8_dynamic_scale_for_fsdp --training.tensor_parallel_degree 2`

this week we are refacting float8 APIs and you may expereince breakage. We float8 to stabalize in the week of 8/5

**Comment by jeromeku** [comment]
@weifengpy 

Thanks for the thorough response!

Can you elaborate on why there is no distinction between tensor parallel (TP) and sequence parallel (SP) in torchtitan?  In the canonical [implemention](https://arxiv.org/abs/2205.05198), these are considered two distinct forms of parallelism as TP splits the weights while SP splits layernorm / drop activations.  Moreover, the `DTensor` API has `RowwiseParallel` / `ColwiseParallel` primitives for TP and a distinct `SequenceParallel` object for SP per the TP / SP [tutorial](https://pytorch.org/tutorials/intermediate/TP_tutorial.html#apply-sequence-parallel-to-layernorm-rmsnorm-layers) as SP can be applied on top of TP.

**Comment by awgu** [comment]
I think it is just a naming thing. The Nvidia paper (https://arxiv.org/abs/2205.05198) called sharding the norm inputs/outputs across the tensor-parallel mesh "sequence parallelism". At the same time, "sequence parallelism" is kind of overloaded since there are also ways to parallelize the inner attention on the sequence dimension, which Nvidia now calls "context parallelism", which is similar to the old "sequence parallelism" paper (https://arxiv.org/abs/2105.13120).

torchtitan probably calls it all TP because it is on the TP mesh dimension. In contrast, context parallelism would require carving out another mesh dimension distinct from TP.

**Comment by tianyu-l** [comment]
@jeromeku 

Happy to answer your question on TP and SP.

In a very narrow sense, yes they are different. Let's use the Llama model in torchtitan as an example. TP splits the weights, and the weights being split are from (1) attention layer and (2) feed forward layer. SP splits activations, and the activations it splits are on dropout and layernorm. So as you can already tell, they are (designed to be) applied on disjoint part of the model forward pass, and can coexist.

This is not a coincidence. In a broader sense, the concept of SP is an improvement to TP in the sense that SP paper (you cited) is actually about (a) TP + (b) splitting dropout / layernorm activations. Merely applying (b) won't give you meaningful benefits (can you see why?). It only makes sense to apply (b) on top of (a). In some sense we can call SP by TP2, or TP+.

Given their close relationship, we'd better have a single name to differentiate from other very different parallelisms (DDP, FSDP, PP, CP, etc.). We ended up using TP because what I explained above (SP is an improvement to the well-known TP and only makes sense when TP is applied).

**Comment by jeromeku** [comment]
@awgu @weifengpy @tianyu-l 

Understood -- many thanks for the responses!

**Comment by zigzagcai** [comment]
> I think it is just a naming thing. The Nvidia paper (https://arxiv.org/abs/2205.05198) called sharding the norm inputs/outputs across the tensor-parallel mesh "sequence parallelism". At the same time, "sequence parallelism" is kind of overloaded since there are also ways to parallelize the inner attention on the sequence dimension, which Nvidia now calls "context parallelism", which is similar to the old "sequence parallelism" paper (https://arxiv.org/abs/2105.13120).
> 
> torchtitan probably calls it all TP because it is on the TP mesh dimension. In contrast, context parallelism would require carving out another mesh dimension distinct from TP.

Thank you for the clarification! It really helps!
